From 8e52e2169b20f7bc3c6c520d4f0c98e5bf06b076 Mon Sep 17 00:00:00 2001
From: Henri Yandell <bayard@apache.org>
Date: Tue, 31 Oct 2006 22:15:00 +0000
Subject: [PATCH] Fixing #LANG-294. The indexOf method did not take into
 account the size variable that limited the amount of the buffer that should
 be looked at.

git-svn-id: https://svn.apache.org/repos/asf/jakarta/commons/proper/lang/trunk@469677 13f79535-47bb-0310-9956-ffa450edef68
---
 .../org/apache/commons/lang/text/StrBuilder.java    |  3 ++-
 .../apache/commons/lang/text/StrBuilderTest.java    | 13 +++++++++++++
 2 files changed, 15 insertions(+), 1 deletion(-)

diff --git a/src/java/org/apache/commons/lang/text/StrBuilder.java b/src/java/org/apache/commons/lang/text/StrBuilder.java
index e202d737259..c7cbfb3e13b 100644
--- a/src/java/org/apache/commons/lang/text/StrBuilder.java
+++ b/src/java/org/apache/commons/lang/text/StrBuilder.java
@@ -1773,8 +1773,9 @@ public int indexOf(String str, int startIndex) {
             return -1;
         }
         char[] thisBuf = buffer;
+        int len = size - strLen + 1;
         outer:
-        for (int i = startIndex; i < thisBuf.length - strLen; i++) {
+        for (int i = startIndex; i < len; i++) {
             for (int j = 0; j < strLen; j++) {
                 if (str.charAt(j) != thisBuf[i + j]) {
                     continue outer;
diff --git a/src/test/org/apache/commons/lang/text/StrBuilderTest.java b/src/test/org/apache/commons/lang/text/StrBuilderTest.java
index f17b1382420..900b922497a 100644
--- a/src/test/org/apache/commons/lang/text/StrBuilderTest.java
+++ b/src/test/org/apache/commons/lang/text/StrBuilderTest.java
@@ -1728,4 +1728,17 @@ public void testToStringBuffer() {
         assertEquals(new StringBuffer("junit").toString(), sb.toStringBuffer().toString());
     }
 
+    //-----------------------------------------------------------------------
+    public void testLang294() {
+        StrBuilder sb = new StrBuilder("\n%BLAH%\nDo more stuff\neven more stuff\n%BLAH%\n");
+        sb.deleteAll("\n%BLAH%");
+        assertEquals("\nDo more stuff\neven more stuff\n", sb.toString()); 
+    }
+
+    public void testIndexOfLang294() {
+        StrBuilder sb = new StrBuilder("onetwothree");
+        sb.deleteFirst("three");
+        assertEquals(-1, sb.indexOf("three"));
+    }
+
 }
